### PR TITLE
Nadam optimizer and test for it added

### PR DESCRIPTION
**Pull request by iikulikov**
Hello @fchollet, this is my first ever PR sorry for any mistakes. 
in my experiments I tried the combination of Adam and NAG from [here](http://cs229.stanford.edu/proj2015/054_report.pdf), and it was better than pure Adam optimizer, today I added it to Keras optimizer module and created this PR. I hope it will be helpful for someone =)

Here is some small experiments which I did, but authors from the report did bigger experiments.

The test passed ok for Theano backend, I don't have TF installed, but if it is necessary I can install it and check.

This is my results comparison for default CIFAR cnn demo from examples folder. (Mini batch = 128, the rest is default).

SGD + NAG:
Using Theano backend.
Using gpu device 0: GeForce GTX 970 (CNMeM is disabled, cuDNN 5004)
X_train shape: (50000, 3, 32, 32)
50000 train samples
10000 test samples
Using real-time data augmentation.
Epoch 1/10
50000/50000 - 22s - loss: 1.8430 - acc: 0.3180 - val_loss: 1.4731 - val_acc: 0.4604
Epoch 2/10
50000/50000 - 19s - loss: 1.4929 - acc: 0.4531 - val_loss: 1.2950 - val_acc: 0.5366
Epoch 3/10
50000/50000 - 19s - loss: 1.3450 - acc: 0.5133 - val_loss: 1.1525 - val_acc: 0.5932
Epoch 4/10
50000/50000 - 19s - loss: 1.2316 - acc: 0.5584 - val_loss: 1.0146 - val_acc: 0.6413
Epoch 5/10
50000/50000 - 15s - loss: 1.1343 - acc: 0.5966 - val_loss: 0.9389 - val_acc: 0.6674
Epoch 6/10
50000/50000 - 12s - loss: 1.0642 - acc: 0.6212 - val_loss: 0.8569 - val_acc: 0.7000
Epoch 7/10
50000/50000 - 12s - loss: 1.0075 - acc: 0.6422 - val_loss: 0.8149 - val_acc: 0.7147
Epoch 8/10
50000/50000 - 12s - loss: 0.9612 - acc: 0.6606 - val_loss: 0.7856 - val_acc: 0.7289
Epoch 9/10
50000/50000 - 12s - loss: 0.9298 - acc: 0.6715 - val_loss: 0.7820 - val_acc: 0.7272
Epoch 10/10
50000/50000 - 12s - loss: 0.8945 - acc: 0.6837 - val_loss: 0.7276 - val_acc: 0.7493

Adam:
Using Theano backend.
Using gpu device 0: GeForce GTX 970 (CNMeM is disabled, cuDNN 5004)
X_train shape: (50000, 3, 32, 32)
50000 train samples
10000 test samples
Using real-time data augmentation.
Epoch 1/10
50000/50000 - 13s - loss: 1.7587 - acc: 0.3508 - val_loss: 1.3012 - val_acc: 0.5314
Epoch 2/10
50000/50000 - 13s - loss: 1.3630 - acc: 0.5087 - val_loss: 1.1107 - val_acc: 0.6094
Epoch 3/10
50000/50000 - 13s - loss: 1.1711 - acc: 0.5831 - val_loss: 0.9536 - val_acc: 0.6626
Epoch 4/10
50000/50000 - 13s - loss: 1.0511 - acc: 0.6275 - val_loss: 0.8502 - val_acc: 0.7025
Epoch 5/10
50000/50000 - 13s - loss: 0.9661 - acc: 0.6589 - val_loss: 0.8260 - val_acc: 0.7083
Epoch 6/10
50000/50000 - 13s - loss: 0.9174 - acc: 0.6780 - val_loss: 0.7416 - val_acc: 0.7409
Epoch 7/10
50000/50000 - 13s - loss: 0.8853 - acc: 0.6883 - val_loss: 0.7118 - val_acc: 0.7529
Epoch 8/10
50000/50000 - 13s - loss: 0.8412 - acc: 0.7022 - val_loss: 0.7058 - val_acc: 0.7527
Epoch 9/10
50000/50000 - 13s - loss: 0.8131 - acc: 0.7156 - val_loss: 0.6746 - val_acc: 0.7623
Epoch 10/10
50000/50000 - 13s - loss: 0.7941 - acc: 0.7191 - val_loss: 0.6801 - val_acc: 0.7622

NAG + Adam (this PR):
Using Theano backend.
Using gpu device 0: GeForce GTX 970 (CNMeM is disabled, cuDNN 5004)
X_train shape: (50000, 3, 32, 32)
50000 train samples
10000 test samples
Using real-time data augmentation.
Epoch 1/10
50000/50000 - 13s - loss: 1.7984 - acc: 0.3485 - val_loss: 1.2916 - val_acc: 0.5318
Epoch 2/10
50000/50000 - 13s - loss: 1.3098 - acc: 0.5289 - val_loss: 1.0351 - val_acc: 0.6300
Epoch 3/10
50000/50000 - 13s - loss: 1.1152 - acc: 0.6049 - val_loss: 0.8808 - val_acc: 0.6840
Epoch 4/10
50000/50000 - 13s - loss: 0.9938 - acc: 0.6498 - val_loss: 0.8003 - val_acc: 0.7159
Epoch 5/10
50000/50000 - 13s - loss: 0.9211 - acc: 0.6755 - val_loss: 0.7352 - val_acc: 0.7422
Epoch 6/10
50000/50000 - 14s - loss: 0.8693 - acc: 0.6952 - val_loss: 0.7321 - val_acc: 0.7469
Epoch 7/10
50000/50000 - 15s - loss: 0.8414 - acc: 0.7066 - val_loss: 0.7250 - val_acc: 0.7552
Epoch 8/10
50000/50000 - 13s - loss: 0.8122 - acc: 0.7186 - val_loss: 0.6938 - val_acc: 0.7627
Epoch 9/10
50000/50000 - 13s - loss: 0.7910 - acc: 0.7261 - val_loss: 0.6349 - val_acc: 0.7838
Epoch 10/10
50000/50000 - 15s - loss: 0.7747 - acc: 0.7319 - val_loss: 0.6148 - val_acc: 0.7902

### CODE REVIEW

**Comment by xingdi-eric-yuan** [comment]
this looks nice

**Comment by the-moliver** [comment]
Nice work! I would suggest you expose the three constants you have hardcoded in these lines so they can be modified by the user:
```python
schedule_decay = 0.004  # Exactly given in [1] and [2]
momentum_cache_t = self.beta_1 * (1. - 0.5 * (K.pow(0.96, t * schedule_decay)))
momentum_cache_t_1 = self.beta_1 * (1. - 0.5 * (K.pow(0.96, (t + 1) * schedule_decay)))
```

**Comment by iikulikov** [comment]
@the-moliver thanks! I also thought about that and when I did more experiments some time ago not with keras, I tried to jiggle this values without any improvements even more, usually I got worse performance.

This constants came from [here](http://cs229.stanford.edu/proj2015/054_report.pdf), page 4, equation 5, authors said that this constants came from earlier Nesterov papers (one of them is from 1983) and he motivated these values also by some deeper research.

If @fchollet ask to expose this variables too, then I will do that. But I think the best is to give meaningful comment in documentation about these values. (I actually will do that in doc-string I guess).

**Comment by fchollet** [comment]
In general we won't merge into Keras algorithms that aren't widely accepted or haven't been covered in a peer-reviewed paper. At the same time, we try to stay on top of things and incorporate the latest advances --as soon as we are confident in their viability.

**Comment by fchollet** [comment]
After external feedback, it seems like this is pretty cool and we'll merge it. Thanks!
